### ISSUE TITLE
Compile State grows after each update

### ISSUE THREAD

**Issue by hybby** [sparkleformation/sparkle_formation#202]
Hello, I'm experiencing an issue where the "Compile State" output of a stack created with Sparkleformation and `sfn-parameters` grows on each successive update.

I wonder where there's an issue where the value is being set with the contents of the previous value somewhere.

Although this isn't causing any problems yet, I'm concerned that ultimately after a number of updates I'm going to hit some sort of resource limit within AWS for the size of an output.

### Version
```
$ sfn --version
sfn - SparkleFormation CLI - [Version: 3.0.20]

$ gem list | grep sfn
sfn (3.0.20)
sfn-parameters (0.2.4)
```

### Template
Simple template to create one SQS queue created as `lib/sparkleformation/templates/template.rb`.

```ruby
compile_time_parameters = {
  description: {
    type: :string,
    default: 'Stack Description'
  }
}

SparkleFormation.new(:test_template, inherit: :base,
                     compile_time_parameters: compile_time_parameters).overrides do
  description state!(:description)

  dynamic!(:sqs_queue, "test_queue") do
    properties do
      queue_name "test_queue"
      delay_seconds 0
      maximum_message_size 262_144
      message_retention_period 345_600
      receive_message_wait_time_seconds 20
      visibility_timeout 30
    end
  end
end
```

### Stack
Simple stack using `sfn-parameters` Stacks Mode created as `stacks/stack.rb`

### Stack creation

```ruby
Configuration.new do
  template 'test_template'

  compile_parameters do
    description 'Initial stack creation'
  end
end
```

### Stack Creation

```
$ sfn create stack
[Sfn]: Callback after_config_create parameters_stacks: starting
[Sfn]: Callback after_config_create parameters_stacks: complete
[Sfn]: Callback template stack_policy: starting
[Sfn]: Compile time parameters: - template: stack
[Sfn]: Description [Initial stack creation]:
[Sfn]: Callback template stack_policy: complete
[Sfn]: SparkleFormation: create
[Sfn]:   -> Name: stack
[Sfn]: Stack runtime parameters: - template: stack
[Sfn]: Environment [testing]:
[Sfn]: Project [infrastructure]:
[Sfn]: Stack Creator [drewjess@drew-laptop.local]:
[Sfn]: Events for Stack: stack
Time                      Resource Logical Id   Resource Status      Resource Status Reason
2017-01-19 15:20:37 UTC   stack                 CREATE_IN_PROGRESS   User Initiated
2017-01-19 15:20:42 UTC   TestQueueSqsQueue     CREATE_IN_PROGRESS
2017-01-19 15:20:43 UTC   TestQueueSqsQueue     CREATE_IN_PROGRESS   Resource creation Initiated
2017-01-19 15:20:44 UTC   TestQueueSqsQueue     CREATE_COMPLETE
2017-01-19 15:20:47 UTC   stack                 CREATE_COMPLETE
[Sfn]: Stack create complete: SUCCESS
[Sfn]: Stack description of stack:
[Sfn]: Outputs for stack: stack
[Sfn]:    Project: infrastructure
[Sfn]:    Compile State: {"stack__description":"Initial stack creation","description":"Initial stack creation"}
[Sfn]:    Environment: testing
[Sfn]:    Stack Creator: drewjess@drew-laptop.local

[Sfn]: Callback after_create stack_policy: starting
[Sfn]: Submitting stack policy documents
[Sfn]: Applying stack policy to stack... complete!
[Sfn]: Stack policy documents successfully submitted!
[Sfn]: Callback after_create stack_policy: complete
```

### Update One
I change the value of `description` in `stacks/stack.rb` from "Initial stack creation" to "Update one" and update the stack:

```
$ sfn update stack
[Sfn]: Callback after_config_update parameters_stacks: starting
[Sfn]: Callback after_config_update parameters_stacks: complete
[Sfn]: Callback template stack_policy: starting
[Sfn]: Compile time parameters: - template: stack
[Sfn]: Description [Update one]:
[Sfn]: Callback template stack_policy: complete
[Sfn]: SparkleFormation: update
[Sfn]:   -> Name: stack Path: /Users/drewjess/git/fd-sfn-dfs/lib/sparkleformation/templates/template.rb
[Sfn]: Stack runtime parameters: - template: stack
[Sfn]: Environment [testing]:
[Sfn]: Project [infrastructure]:
[Sfn]: Stack Creator [drewjess@drew-laptop.local]:
[Sfn]: Pre-update resource planning report:

  Update plan for: stack
    No resource lifecycle changes detected!

[Sfn]: Apply this stack update? (Y/N): y
[Sfn]: Callback before_update stack_policy: starting
[Sfn]: Callback before_update stack_policy: complete
[Sfn]: Events for Stack: stack
Time                      Resource Logical Id   Resource Status      Resource Status Reason
2017-01-19 15:21:24 UTC   stack                 UPDATE_IN_PROGRESS   User Initiated
2017-01-19 15:21:29 UTC   stack                 UPDATE_COMPLETE_CLEANUP_IN_PROGRESS
2017-01-19 15:21:30 UTC   stack                 UPDATE_COMPLETE
[Sfn]: Stack update complete: SUCCESS
[Sfn]: Stack description of stack:
[Sfn]: Outputs for stack: stack
[Sfn]:    Project: infrastructure
[Sfn]:    Compile State: {"stack__description":"Update one","stack__stack":{"stack__description":"Initial stack creation","description":"Initial stack creation"},"description":"Update one"}
[Sfn]:    Environment: testing
[Sfn]:    Stack Creator: drewjess@drew-laptop.local

[Sfn]: Callback after_update stack_policy: starting
[Sfn]: Submitting stack policy documents
[Sfn]: Applying stack policy to stack... complete!
[Sfn]: Stack policy documents successfully submitted!
[Sfn]: Callback after_update stack_policy: complete
```

### Update Two
I change the value of `description` in `stacks/stack.rb` from "Update one" to "Update two" and update the stack:

```
$ sfn update stack
[Sfn]: Callback after_config_update parameters_stacks: starting
[Sfn]: Callback after_config_update parameters_stacks: complete
[Sfn]: Callback template stack_policy: starting
[Sfn]: Compile time parameters: - template: stack
[Sfn]: Description [Update two]:
[Sfn]: Callback template stack_policy: complete
[Sfn]: SparkleFormation: update
[Sfn]:   -> Name: stack Path: /Users/drewjess/git/fd-sfn-dfs/lib/sparkleformation/templates/template.rb
[Sfn]: Stack runtime parameters: - template: stack
[Sfn]: Environment [testing]:
[Sfn]: Project [infrastructure]:
[Sfn]: Stack Creator [drewjess@drew-laptop.local]:
[Sfn]: Pre-update resource planning report:

  Update plan for: stack
    No resource lifecycle changes detected!

[Sfn]: Apply this stack update? (Y/N): y
[Sfn]: Callback before_update stack_policy: starting
[Sfn]: Callback before_update stack_policy: complete
[Sfn]: Events for Stack: stack
Time                      Resource Logical Id   Resource Status      Resource Status Reason
2017-01-19 15:22:17 UTC   stack                 UPDATE_IN_PROGRESS   User Initiated
2017-01-19 15:22:21 UTC   stack                 UPDATE_COMPLETE_CLEANUP_IN_PROGRESS
2017-01-19 15:22:23 UTC   stack                 UPDATE_COMPLETE
[Sfn]: Stack update complete: SUCCESS
[Sfn]: Stack description of stack:
[Sfn]: Outputs for stack: stack
[Sfn]:    Project: infrastructure
[Sfn]:    Compile State: {"stack__description":"Update two","stack__stack":{"stack__description":"Update one","stack__stack":{"stack__description":"Initial stack creation","description":"Initial stack creation"},"description":"Update one"},"description":"Update two"}
[Sfn]:    Environment: testing
[Sfn]:    Stack Creator: drewjess@drew-laptop.local

[Sfn]: Callback after_update stack_policy: starting
[Sfn]: Submitting stack policy documents
[Sfn]: Applying stack policy to stack... complete!
[Sfn]: Stack policy documents successfully submitted!
[Sfn]: Callback after_update stack_policy: complete
```

### Update Three
I change the value of `description` in `stacks/stack.rb` from "Update two" to "Update three" and update the stack:

```
$ sfn update stack
[Sfn]: Callback after_config_update parameters_stacks: starting
[Sfn]: Callback after_config_update parameters_stacks: complete
[Sfn]: Callback template stack_policy: starting
[Sfn]: Compile time parameters: - template: stack
[Sfn]: Description [Update three]:
[Sfn]: Callback template stack_policy: complete
[Sfn]: SparkleFormation: update
[Sfn]:   -> Name: stack Path: /Users/drewjess/git/fd-sfn-dfs/lib/sparkleformation/templates/template.rb
[Sfn]: Stack runtime parameters: - template: stack
[Sfn]: Environment [testing]:
[Sfn]: Project [infrastructure]:
[Sfn]: Stack Creator [drewjess@drew-laptop.local]:
[Sfn]: Pre-update resource planning report:

  Update plan for: stack
    No resource lifecycle changes detected!

[Sfn]: Apply this stack update? (Y/N): y
[Sfn]: Callback before_update stack_policy: starting
[Sfn]: Callback before_update stack_policy: complete
[Sfn]: Events for Stack: stack
Time                      Resource Logical Id   Resource Status      Resource Status Reason
2017-01-19 15:23:06 UTC   stack                 UPDATE_IN_PROGRESS   User Initiated
2017-01-19 15:23:11 UTC   stack                 UPDATE_COMPLETE_CLEANUP_IN_PROGRESS
2017-01-19 15:23:12 UTC   stack                 UPDATE_COMPLETE
[Sfn]: Stack update complete: SUCCESS
[Sfn]: Stack description of stack:
[Sfn]: Outputs for stack: stack
[Sfn]:    Project: infrastructure
[Sfn]:    Compile State: {"stack__description":"Update three","stack__stack":{"stack__description":"Update two","stack__stack":{"stack__description":"Update one","stack__stack":{"stack__description":"Initial stack creation","description":"Initial stack creation"},"description":"Update one"},"description":"Update two"},"description":"Update three"}
[Sfn]:    Environment: testing
[Sfn]:    Stack Creator: drewjess@drew-laptop.local

[Sfn]: Callback after_update stack_policy: starting
[Sfn]: Submitting stack policy documents
[Sfn]: Applying stack policy to stack... complete!
[Sfn]: Stack policy documents successfully submitted!
[Sfn]: Callback after_update stack_policy: complete
```

### Summary
As you can see, the size of the "Compile State" output grows upon each update:

```
[Sfn]:    Compile State: {"stack__description":"Initial stack creation","description":"Initial stack creation"}
```

```
[Sfn]:    Compile State: {"stack__description":"Update one","stack__stack":{"stack__description":"Initial stack creation","description":"Initial stack creation"},"description":"Update one"}
```

```
[Sfn]:    Compile State: {"stack__description":"Update two","stack__stack":{"stack__description":"Update one","stack__stack":{"stack__description":"Initial stack creation","description":"Initial stack creation"},"description":"Update one"},"description":"Update two"}
```

```
[Sfn]:    Compile State: {"stack__description":"Update three","stack__stack":{"stack__description":"Update two","stack__stack":{"stack__description":"Update one","stack__stack":{"stack__description":"Initial stack creation","description":"Initial stack creation"},"description":"Update one"},"description":"Update two"},"description":"Update three"}
```

As mentioned, this isn't causing any issues at the moment, but I'm concerned that ultimately after a number of updates I'm going to hit some sort of resource limit within AWS for the size of an output.  

Operationally, I can imagine running hundreds of stack updates over the lifetime of a stack.  The produced CloudFormation template may become overly bulky after this many updates.

**Comment by hybby** [comment]
Out of curiosity, I pushed this to the (undocumented) limit.  If the value "Compile State" becomes greater than 51200 characters, we get the following error: 

```
ERROR: Miasma::Error::ApiError::RequestError: Bad Request - ValidationError: 1 validation error detected: Value '{"AWSTemplateFormatVersion":"2010-09-09","Parameters":{"Environment":{"Type":"String","Default":"testing"},"Project":{"Type":"String","Default":"infrastructure"},"StackCreator":{"Type":"String","Default":"drewjess@drew-laptop.local"}},"Outputs":{"Environment":{"Description":"Environment parameter","Value":{"Ref":"Environment"}},"Project":{"Description":"Project parameter","Value":{"Ref":"Project"}},"StackCreator":{"Description":"Stack creator parameter","Value":{"Ref":"StackCreator"}},"CompileState":{"Value":
...snip...
at 'templateBody' failed to satisfy constraint: Member must have length less than or equal to 51200
```

It's unlikely I'm going to hit this any time soon, but shows that at some point there *may* indeed be issues... :)

**Comment by chrisroberts** [comment]
Fixed via #241 and will be in the next sfn release. Thanks for the report!
